plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completion/Completer/FilePath.hs

```diff
@@ -0,0 +1,190 @@
+        pure $ mkCompletionItem (completionRange prefInfo) fullDirPath fullDirPath
+    )
+
+{- Note [Using correct file path separators]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  Since cabal files only allow for posix style file paths
+  we need to be careful to use the correct path separators
+  whenever we work with file paths in cabal files.
+
+  Thus we are using two different kinds of imports.
+  We use "FP" for platform-compatible file paths with which
+  we can query files independently of the platform.
+  We use "Posix" for the posix syntax paths which need to
+  be used for file path completions to be written to the cabal file.
+-}
+
+{- | Information used to query and build file path/directory completions.
+
+  Note that partialFileName combined with partialFileDir results in
+  the original prefix.
+
+  Example:
+  On the written filepath: @dir1/fi@ the
+  resulting PathCompletionInfo would be:
+
+  @
+    partialFileName = "fi"
+    partialFileDir  = "dir1/dir2/fi"
+    ...
+  @
+-}
+data PathCompletionInfo = PathCompletionInfo
+  { partialFileName :: T.Text
+  -- ^ partly written start of next part of path
+  , partialFileDir  :: FilePath
+  -- ^ written part of path
+  , workingDir      :: FilePath
+  -- ^ current working directory of the handled file
+  }
+  deriving (Eq, Show, Read)
+
+{- | Takes a path completion info and returns the list of files
+  in the directory that match the path completion info.
+-}
+listFileCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listFileCompletions recorder complInfo = do
+  let complDir = mkCompletionDirectory complInfo
+  try (evaluate =<< listDirectory complDir) >>= \case
+    Right dirs -> do
+      fixedDirs <-
+        mapM
+          ( \d -> do
+              isDir <- doesDirectoryExist $ mkDirFromCWD complInfo d
+              pure $ if isDir then Posix.addTrailingPathSeparator d else d
+          )
+          dirs
+      pure fixedDirs
+    Left (err :: IOError) -> do
+      logWith recorder Warning $ LogFilePathCompleterIOError complDir err
+      pure []
+
+{- | Returns a list of all (and only) directories in the
+  directory described by path completion info.
+-}
+listDirectoryCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listDirectoryCompletions recorder complInfo = do
+  filepaths <- listFileCompletions recorder complInfo
+  filterM (doesDirectoryExist . mkDirFromCWD complInfo) filepaths
+
+pathCompletionInfoFromCabalPrefixInfo :: CabalPrefixInfo -> PathCompletionInfo
+pathCompletionInfoFromCabalPrefixInfo ctx =
+  PathCompletionInfo
+    { partialFileName = dirNamePrefix
+    , partialFileDir = Posix.addTrailingPathSeparator $ Posix.takeDirectory prefix
+    , workingDir = dir
+    }
+ where
+  prefix = T.unpack $ completionPrefix ctx
+  dirNamePrefix = T.pack $ Posix.takeFileName prefix
+  dir = completionWorkingDir ctx
+
+{- | Returns the directory, the currently handled cabal file is in.
```

Comment:
Is this right? It looks to me like it's going to return the directory that the partial completion is pointing to instead?

Reply:
maybe `dirNamePrefix` is not the best name but i think it returns the last part of the written path which is either a partially written directory name or file name.
The intention is that on a written path like `src/dire` i have `src/` as the partialFileDir and `dire` as the partialFileName.

Reply:
Right, but that's not the location of the _cabal file_, is it? I thought that was the working directory (if anything). TBH I would probably try and abstract away the cabal file here, and just talk about working directories etc.

Reply:
Ah, I missed that you were talking about the documentation here :D
I will update it.